### PR TITLE
Clean-up IOperation tree representing tuple literal conversions.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #22548 and #22229.

There are several issues with IOperation tree when it comes to conversions applied to tuple literals:
1. When conversion is explicitly specified in code, it is translated to conversions on individual
elements of the literal and these BoundConversion nodes are also marked as explicitly
specified in code (which makes sense). However, the syntax associated with a BoundConversion
node on top of an element matches the syntax of the element (which also makes sense).
The problem is that both the conversion node and the element appear as “explicit” in
IOperation tree, which violates one of the tree invariants – at most one “explicit” node for
a given syntax node.
2. When a tuple literal is converted, compiler creates a BoundConvertedTupleLiteral node,
which has the target tuple type as its Type, but the BoundConversion node on top of the
BoundConvertedTupleLiteral captures conversion information (kind, etc.) for conversion
From tuple’s natural type to the target type. So, often, we end up with both nodes having
The same Type, but the conversion says it is not an identity conversion. This shape helps
SemanticModel to provide accurate information about types and conversions, but the
IOperation tree looks very confusing.
3. When conversion is explicitly specified in code, the BoundConversion node from the
previous item is marked as explicitly specified in code, but it shares syntax node with
BoundConvertedTupleLiteral. Both nodes appear as “explicit” in IOperation tree, which
violates the same invariant.
4. When conversion is explicitly specified in code, compiler adds yet another explicit
Identity conversion on top of the BoundConversion node from the previous item.
It is needed for proper SemanticModel behavior. Even when it uses different syntax,
the IOperation tree looks very confusing, there are two conversions on top of each other
and it is not clear why intermediate conversion node is needed.

In order to address all these issue the following changes are made to IOperation factory:
- If conversion node is associated with the same syntax as its operand, the conversion
is marked as isImplicit.
- If Type of BoundConvertedTupleLiteral matches the Type of BoundConversion node
on top of it the conversion node is not added to the IOperation tree. Instead a new property
is added to ITupleOperation interface, called NaturalType. The property exposes the natural
type of the tuple literal and makes it easy for consumers to figure out if a conversion from
a natural type of a tuple literal took place.
-  If Type of BoundConvertedTupleLiteral doesn’t match the Type of BoundConversion node
on top of it and there is an identity conversion node on top of the BoundConversion node,
the intermediate conversion node is not added to the IOperation tree, its conversion
information is used for the top-most conversion node (instead of saying that it is identity).

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review IOperation fix.

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review IOperation fix.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 1]
```diff
@@ -564,7 +565,40 @@ Namespace Microsoft.CodeAnalysis.Operations
...
+                Dim nestedConversionOperandAndMethod As (Operand As BoundExpression, Method As MethodSymbol) = GetConversionInfo(nestedConversion)
+                Dim nestedOperand As BoundExpression = nestedConversionOperandAndMethod.Operand
+
+                If nestedConversion.Syntax Is nestedOperand.Syntax AndAlso nestedConversion.ExplicitCastInCode AndAlso
+                   nestedOperand.Kind = BoundKind.ConvertedTupleLiteral AndAlso
+                   nestedConversion.Type <> nestedOperand.Type Then
+                    ' Let's erase the nested conversion, this is an artificial conversion added on top of BoundConvertedTupleLiteral
+                    ' in Binder.ReclassifyTupleLiteral.
+                    ' We need to use conversion information from the nested conversion because that Is where the real conversion 
+                    ' information Is stored.
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: `is` rather than `Is` (two locations).

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ITupleOperation.cs [hunk 2]
```diff
@@ -22,6 +22,13 @@ public interface ITupleOperation : IOperation
         /// Tuple elements.
         /// </summary>
         ImmutableArray<IOperation> Elements { get; }
+
+        /// <summary>
+        /// Natural type of the tuple, or null if tuple doesn't have a natural type.
+        /// Natural type can be different from <see cref="IOperation.Type"/> depending on the
+        /// conversion context, in which the tupe is used.
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ITupleOperation.cs]: tupe -> tuple

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs [hunk 3]
```diff
@@ -186,6 +188,83 @@ static void Main()
...
+        Console.WriteLine(t);
+    }
+}
+";
+            string expectedOperationTree = @"
+IVariableInitializerOperation (OperationKind.VariableInitializer, Type: null) (Syntax: '= (a, b)')
+  IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: (System.Int64, System.Int64), IsImplicit) (Syntax: '(a, b)')
+    Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+    Operand: 
+      ITupleOperation (OperationKind.Tuple, Type: (System.Int64 a, System.Int64 b)) (Syntax: '(a, b)')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs]: Just to make sure that I'm understanding the conversion here, the actual thing being converted here is that the names of the tuple elements are being erased, correct? As both elements are still longs.

**Review by 333fred — APPROVED** [review]
Assuming my conversion comment was correct, LGTM

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs]: Your understanding feels correct.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 4]
```diff
@@ -4825,11 +4825,19 @@ public LazyTryStatement(Lazy<IBlockOperation> body, Lazy<ImmutableArray<ICatchCl
...
+        protected BaseTupleExpression(SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, ITypeSymbol naturalType, Optional<object> constantValue, bool isImplicit) :
                     base(OperationKind.Tuple, semanticModel, syntax, type, constantValue, isImplicit)
         {
+            NaturalType = naturalType;
         }
 
+        /// <summary>
+        /// Natural type of the tuple, or null if tuple doesn't have a natural type.
+        /// Natural type can be different from <see cref="IOperation.Type"/> depending on the
+        /// conversion context, in which the tuple is used.
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Nit: unnecessary comma

**Review by mavasani — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks!

**Comment by AlekseyTs** [comment]
Test ubuntu_16_debug_prtest please

**Comment by AlekseyTs** [comment]
https://ci.dot.net/job/dotnet_roslyn/job/post-dev15.5-contrib/job/windows_debug_unit32_prtest/354/

CLR test failure (only matches the first, see "Latest Test Result" link for the rest)

Roslyn.VisualStudio.Next.UnitTests.Remote.ServiceHubServicesTests.TestRemoteHostConnect [FAIL]
Indication 1

**Comment by AlekseyTs** [comment]
Test windows_debug_unit32_prtest please

**Pull request merged**